README.md
[![preroll on crates.io](https://img.shields.io/crates/v/preroll)](https://crates.io/crates/preroll) [![Documentation (latest release)](https://docs.rs/preroll/badge.svg)](https://docs.rs/preroll/)

# preroll

Easy boilerplate utilities for Rust http services which use [async-std][], [Tide][], [Surf][], and friends.

Allows for service setup with feature-configured built-ins for maximum service consistency with low developer overhead,
and for easily integration testing the service without using a live network.

**Scroll to the bottom for API Reference**

### Example

```rust
use std::sync::Arc;

use tide::{Request, Route};

struct AppState {
    greeting: &'static str,
}

type AppRequest = Request<Arc<AppState>>;

async fn setup_app_state() -> preroll::SetupResult<AppState> {
    Ok(AppState {
        greeting: "Hello World!",
    })
}

fn setup_routes(mut server: Route<'_, Arc<AppState>>) {
    server
        .at("hello-world")
        .get(|req: AppRequest| async move {
            Ok(req.state().greeting)
        });
}

// The "magic" happens here!
preroll::main!("hello-world", setup_app_state, setup_routes);
```

### Features

- Boilerplate `main` setup via [`preroll::main!`][], with optional features automatically configured.
- A [`preroll::prelude::*;`][] with all extension traits.
- Response logging with many details.
- Automatic JSON responses for errors in the form of [`JsonError`][].
- [Test utils][] with easy mock client setup.

### Optional features
Add-on features must be enabled via cargo features, e.g.

```toml
[dependencies.preroll]
version = "0.5"
features = ["honeycomb", "postgres"]
```

#### List of optional add-on features:
- `"honeycomb"`: Enables tracing to [honeycomb.io].
    - Env variable `HONEYCOMBIO_WRITE_KEY` (required).
    - Env variable `TRACELEVEL`, sets the tracing level filter, defaults to `info`.
    - Writes to a dataset named `{service_name}-{environment}`.
        - `service_name` is from `preroll::main!("service_name", ...)`.
        - `environment` is from `ENVIRONMENT`, or defaults to `"development"`.
- `"lambda-http"`: Changes the HTTP listener to connect to an AWS Lambda execution environment.
    - Is no longer reachable as a regular http server, but accepts http lambda requests as if it were one.
    - Some environment variables, such as `PORT`, are disregarded.
    - If the `"honeycomb"` feature is enabled, trace events are written to stdout, and must be collected via
        a layer provided by Honeycomb. See: https://docs.honeycomb.io/getting-data-in/integrations/aws/aws-lambda/
- `"postgres"`: Enables a postgres connection pool with transactions.
    - Env variable `PGURL`, which should be a properly formatted `postgres://` database url.
        - Defaults to `"postgres://localhost/{service_name}"` (default postgres port).
        - `service_name` is from `preroll::main!("service_name", ...)`.
    - Env variable `PGMAXCONNECTIONS`, default 5 connections.
    - Env variable `PGMAXLIFETIME`, default `30` (minutes).
    - Enables [`PostgresRequestExt`][prelude::PostgresRequestExt] and [`test_utils::create_client_and_postgres`][].

#### List of other optional features:
- `"panic-on-error"`: Makes the response logger [panic][] on error rather than log.
    - Do not use in production. Prevents `--release` compilation.

### General Environment Settings
The following environment variables are read during `preroll::main!`:
- `ENVIRONMENT`: If this starts with `prod`, load the production-mode JSON logger, avoid `.env`.
- `FORCE_DOTENV`: Override production-mode, force-load environment from `.env`.
- `HOST`: Sets the hostname that this service will listen on. Defaults to `"127.0.0.1"`.
- `LOGLEVEL`: Set the logger's level filter, defaults to `info` in production-mode, `debug` in development-mode.
- `PORT`: Sets the port that this service will listen on. Defaults to `8080`.

### Note:

This crate is intentionally somewhat prescriptive in how it templates a service and the interaction with
add-on features such as Postgres (via [SQLx][]).

[`preroll::main!`]: https://docs.rs/preroll/0.8.0/preroll/macro.main.html
[`preroll::prelude::*;`]: https://docs.rs/preroll/0.8.0/preroll/prelude/index.html
[`JsonError`]: https://docs.rs/preroll/0.8.0/preroll/struct.JsonError.html
[async-std]: https://async.rs/
[honeycomb.io]: https://www.honeycomb.io/
[SQLx]: https://github.com/launchbadge/sqlx#sqlx
[Surf]: https://github.com/http-rs/surf#surf
[Test utils]: https://docs.rs/preroll/0.8.0/preroll/test_utils/index.html
[Tide]: https://github.com/http-rs/tide#tide

## API Reference

[API Reference on Docs.rs](https://docs.rs/preroll/0.9.0/preroll/#modules)

## License

Licensed under the [BlueOak Model License 1.0.0](LICENSE.md) — _[Contributions via DCO 1.1](contributing.md#developers-certificate-of-origin)_


Changelog.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.10.1]

- `x-clacks-overhead` header added to maintain feature parity with boltzmann
- `clap` feature of the `petname` dependency disabled to improve build times
## [0.10.0]

### Dependencies

- Update to preroll 0.10
- Update to tracing-subscriber 0.3

## [0.9.0] - 2021-10-28

This release changes test database naming inference and `/monitor/ping` output.

### Changes

- Change: `/monitor/ping` now randomly generates a human-readable name at start-up, and will reply with that for the process duration.
- Change: the default postgres test database name is now `{crate_name}-test`.

### Additions

- Support for overriding `host`, `name` and `port` for the postgres test database.

## [0.8.5] - 2021-09-20

### Changes

This release comes with a minor change to the debug (development) mode defaults.

- Change: `LOGLEVEL` defaults to `info` rather than `debug` in debug mode.

## [0.8.4] - 2021-08-26

### Dependencies
- Update to surf 2.3
- Update to env_logger 0.9

## [0.8.3] - 2021-07-19

### Improvements
- The exported test assertion helpers now use `#[track_caller]` to have better error location / backtrace output.
    - See https://doc.rust-lang.org/reference/attributes/codegen.html#the-track_caller-attribute

## [0.8.2] - 2021-07-12

### Fixes
- `lambda-http`: This feature flag now actually works. (Oops.)

## [0.8.1] - 2021-07-12

### Fixes
- `lambda-http`: No longer disables the logging middleware.

## [0.8.0] - 2021-06-15

### Additions
- New `"lambda-http"` feature, which changes the HTTP listener to connect to an AWS Lambda execution environment.
    - Is no longer reachable as a regular http server, but accepts http lambda requests as if it were one.
    - Some environment variables, such as `PORT`, are disregarded.
    - If the `"honeycomb"` feature is enabled, trace events are written to stdout, and must be collected via
        a layer provided by Honeycomb. See: https://docs.honeycomb.io/getting-data-in/integrations/aws/aws-lambda/

## [0.7.0] - 2021-05-19

### Changes
- The default backend for the included [Surf][] http client has changed from libcurl (via [Isahc][]) to [http-client's async-h1 client][].
- `honeycomb`: Environemtn variables now start with `HONEYCOMB_` rather than `HONEYCOMBIO_`.
    - Consistency with other honeycomb SDK's.
- `honeycomb`: `HONEYCOMBIO_WRITE_KEY` is now `HONEYCOMB_WRITEKEY` (required).
    - Consistency with other honeycomb SDK's.

### Additions
- `honeycomb`: Added `HONEYCOMB_SAMPLE_RATE` environment var usage.

[http-client's async-h1 client]: https://github.com/http-rs/http-client/tree/main/src/h1
[Isahc]: https://github.com/sagebind/isahc
[Surf]: https://github.com/http-rs/surf

## [0.6.0] - 2021-04-15

### Additions
- `postgres`: Added a `PGMAXLIFETIME` environemnt variable, set in minutes.

### Dependencies
- `honeycomb`: Switched back to `tracing-honeycomb` and upgraded to `0.3`.
   - This allows patch upgrades to be picked up via `cargo update`, and is otherwise identical to `0.2.1-eaze.7`.

## [0.5.7] - 2021-04-05

- `honeycomb`: Added events for generic Request/Response http properties.
- Changed the ordering of json log fields to be better for plaintext viewers.
- Fixed hostname resolution in `/monitor/status`.

## [0.5.6] - 2021-03-29

- Dependency upgrade to eaze-tracing-honeycomb 0.2.1-eaze.7
    - `honeycomb`: This fixes a deadlock. See [this commit] for details.

[this commit]: https://github.com/eaze/tracing-honeycomb/commit/9dd18b55ea96b95ce76d0051dbcbd085b7e7f2f1

## [0.5.5] - 2021-03-24

- `honeycomb`: Fixed timestamps to correctly be from spans/events.
- Dependency upgrade to eaze-tracing-honeycomb 0.2.1-eaze.6

## [0.5.4] - 2021-03-24

- `honeycomb`: Sub-millisecond "duration_ms" using f64.
- Dependency upgrade to eaze-tracing-honeycomb 0.2.1-eaze.5

## [0.5.3] - 2021-03-22

- Enables Postgres trace events.
- Dependency upgrade to tide-sqlx 0.6.0

## [0.5.2] - 2021-03-19

- Dependency upgrade to eaze-tracing-honeycomb 0.2.1-eaze.4

## [0.5.1] - 2021-03-09

- Reduce dependency footprint from async-std.
    - Avoids wasm-related features.

## [0.5.0] - 2021-03-09

- Dependency upgrade to sqlx 0.5 & tide-sqlx 0.5 for the `postgres` feature.

## [0.4.3] - 2021-02-22

- Fixes `X-Honeycomb-Trace` header parsing.
    - Under the hood, this meant a switch to [the Eaze fork of tracing-honeycomb](https://github.com/eaze/tracing-honeycomb).

## [0.4.2] - 2021-02-16

- Avoid causing http errors on invalid `X-Honeycomb-Id` and `X-Request-Id`.
    - Instead, these get logged.

## [0.4.1] - 2021-02-02

- Updated the `"honeycomb"` feature to respect the `"HONEYCOMBIO_DATASET"` environment variable when possible.

## [0.4.0] - 2021-02-01

The same as 0.3 but with a forgotten update to Tide 0.16.

- Dependency upgrade to Tide 0.14 & tide-sqlx 0.4.

## [0.3.0] - 2021-02-01

- Removed the `"custom_middleware"` feature, is now automatically part of the optional arguments for `preroll::main!`.
- Changed `preroll::main!` to accept `VariadicRoutes` instead of variable macro arguments.
- Changed `preroll::test_utils::create_client` to accept `VariadicRoutes`, enabled api versioned testing.
- Changed assertions to always unwrap and panic immediately rather than return a `Result`.
- Changed assertions to accept `AsMut<http_types::Response>` rather than `surf::Response`.
- Added more test assertion helpers to `preroll::test_utils`.
- Added `/monitor/status` built-in endpoint to expose useful status information.
- Updated all documentation with major additions, overhauls, and examples.
- Improved the standalone example / test subproject.
- Improved internal CI.

## [0.2.0] - 2021-01-12

- Changed the api mounting point to be versioned - e.g. `/api/v1`.

src/setup.rs
//! The setup components which comprise `preroll::main!`.
//!
//! These are exposed in the event they need to be used more manually, but use is discouraged.
//! Prefer using `preroll::main!` whenever possible.

use std::env;
use std::future::Future;
use std::sync::Arc;

use cfg_if::cfg_if;
use tide::{Request, Server};

pub use async_std::task::block_on;

use crate::builtins::monitor::setup_monitor;

cfg_if! {
    if #[cfg(feature = "honeycomb")] {
        use tracing_honeycomb::new_blackhole_telemetry_layer;
        use tracing_subscriber::filter::LevelFilter;
        use tracing_subscriber::prelude::*;
        use tracing_subscriber::Registry;

        use crate::middleware::TraceMiddleware;
    }
}

cfg_if! {
    if #[cfg(feature = "postgres")] {
        use std::time::Duration;

        use sqlx::postgres::{PgConnectOptions, PgPoolOptions};
        use sqlx::ConnectOptions;

        use crate::middleware::PostgresMiddleware;
    }
}

cfg_if! {
    if #[cfg(feature = "lambda-http")] {
        use tide_lambda_listener::LambdaListener;
    } else {
        use tide::listener::Listener;
    }
}

use crate::logging::{log_format_json, log_format_pretty};
use crate::middleware::{
    ClacksMiddleware, JsonErrorMiddleware, LogMiddleware, RequestIdMiddleware,
};
use crate::VariadicRoutes;

/// The result type which is expected from functions passed to `preroll::main!`,
/// and used in the return of `setup`'s functions.
///
/// This is a `color_eyre::eyre::Result<T>`.
pub type Result<T> = color_eyre::eyre::Result<T>;

pub async fn setup<AppState, StateFn, StateFnFuture, ServerFn, ServerFnFuture>(
    service_name: &'static str,
    state_setup: StateFn,
    server_setup: ServerFn,
    routes_setups: impl Into<VariadicRoutes<AppState>>,
) -> Result<()>
where
    AppState: Send + Sync + 'static,
    StateFn: Fn() -> StateFnFuture,
    StateFnFuture: Future<Output = Result<AppState>>,
    ServerFn: Fn(Server<Arc<AppState>>) -> ServerFnFuture,
    ServerFnFuture: Future<Output = Result<Server<Arc<AppState>>>>,
{
    initial_setup(service_name)?;

    let state = state_setup().await?;

    let (mut base_server, server) = setup_server(service_name, state).await?;

    let mut server = server_setup(server).await?;

    let mut version = 1;
    for routes_fn in routes_setups.into().routes {
        routes_fn(server.at(&format!("/api/v{}", version)));
        version += 1;
    }

    #[cfg(debug_assertions)]
    server.at("/internal-error").get(get_internal_error);

    base_server.at("/").nest(server);
    start_server(base_server).await?;

    Ok(())
}

#[cfg(debug_assertions)]
async fn get_internal_error<AppState>(_req: Request<Arc<AppState>>) -> tide::Result<&'static str>
where
    AppState: Send + Sync + 'static,
{
    Err(tide::Error::from_str(
        500,
        "Intentional Server Error from GET /internal-error",
    ))
}

#[cfg_attr(not(feature = "honeycomb"), allow(unused_variables))]
pub fn initial_setup(service_name: &'static str) -> Result<()> {
    color_eyre::install()?;

    let log_level = env::var("LOGLEVEL")
        .map(|v| v.parse().expect("LOGLEVEL must be a valid log level."))
        .unwrap_or(log::LevelFilter::Info);

    if env::var("FORCE_DOTENV").is_ok() || env::var("DEBUG_DOTENV").is_ok() {
        dotenv::dotenv().ok();
    }

    let environment = env::var("ENVIRONMENT").unwrap_or_else(|_| "development".to_string());

    // Logging
    if environment.starts_with("prod") {
        env_logger::builder()
            .format(log_format_json)
            .filter_level(log_level)
            .write_style(env_logger::WriteStyle::Never)
            .try_init()?;
    } else {
        // Development
        dotenv::dotenv().ok();

        env_logger::builder()
            .format(log_format_pretty)
            .filter_level(log_level)
            .try_init()?;
    }

    log::info!("Logger started - level: {}", log_level);

    // Tracing (Honeycomb)
    #[cfg(feature = "honeycomb")]
    {
        let trace_filter: LevelFilter = env::var("TRACELEVEL")
            .map(|v| v.parse())
            .unwrap_or(Ok(LevelFilter::INFO))?;

        if let Ok(api_key) = env::var("HONEYCOMB_WRITEKEY") {
            let maybe_sample_rate = env::var("HONEYCOMB_SAMPLE_RATE");

            #[cfg(feature = "lambda-http")]
            let telemetry_layer = {
                // Just to avoid unused warnings.
                // In this setup the environemnt's consumer will have to have this.
                drop(api_key);

                let telemetry_builder = tracing_honeycomb::Builder::new_stdout(service_name);

                if let Ok(sample_rate) = maybe_sample_rate.unwrap_or_default().parse() {
                    telemetry_builder.with_trace_sampling(sample_rate /* u32 */)
                } else {
                    telemetry_builder
                }
                .build()
            };

            #[cfg(not(feature = "lambda-http"))]
            let telemetry_layer = {
                let dataset = env::var("HONEYCOMB_DATASET")
                    .unwrap_or_else(|_| format!("{}-{}", service_name, environment));

                let api_host = env::var("HONEYCOMB_API_HOST")
                    .unwrap_or_else(|_| "https://api.honeycomb.io/".to_string());

                let honeycomb_config = libhoney::Config {
                    options: libhoney::client::Options {
                        api_key,
                        api_host,
                        dataset,
                        ..libhoney::client::Options::default()
                    },
                    transmission_options: libhoney::transmission::Options::default(),
                };

                let telemetry_builder =
                    tracing_honeycomb::Builder::new_libhoney(service_name, honeycomb_config);

                if let Ok(sample_rate) = maybe_sample_rate.unwrap_or_default().parse() {
                    telemetry_builder.with_trace_sampling(sample_rate /* u32 */)
                } else {
                    telemetry_builder
                }
                .build()
            };

            let subscriber = Registry::default()
                .with(trace_filter) // filter out low-level debug tracing
                // .with(tracing_subscriber::fmt::Layer::default()) // log to stdout
                .with(telemetry_layer); // publish to honeycomb backend

            tracing::subscriber::set_global_default(subscriber)?;

            log::info!("Honeycomb Tracing enabled - filter: {}", trace_filter);
        } else {
            let telemetry_layer = new_blackhole_telemetry_layer();

            let subscriber = Registry::default()
                .with(trace_filter) // filter out low-level debug tracing
                // .with(tracing_subscriber::fmt::Layer::default()) // log to stdout
                .with(telemetry_layer); // publish to honeycomb backend

            tracing::subscriber::set_global_default(subscriber)?;

            log::info!("Honeycomb Tracing off");
        }
    }

    Ok(())
}

#[cfg_attr(not(feature = "postgres"), allow(unused_variables))]
pub async fn setup_server<State>(
    service_name: &'static str,
    state: State,
) -> Result<(Server<Arc<()>>, Server<Arc<State>>)>
where
    State: Send + Sync + 'static,
{
    let mut base_server = tide::with_state(Arc::new(()));
    base_server.with(ClacksMiddleware::new());

    // Set handlers for /monitor/ping, etc.
    //
    // These are intentionally excluded from logging/tracing middleware.
    setup_monitor(service_name, &mut base_server);

    let mut server = tide::with_state(Arc::new(state));
    server.with(ClacksMiddleware::new());
    server.with(RequestIdMiddleware::new());
    server.with(LogMiddleware::new());
    server.with(JsonErrorMiddleware::new());

    #[cfg(feature = "honeycomb")]
    server.with(TraceMiddleware::new());

    // Postgres
    #[cfg(feature = "postgres")]
    {
        let max_connections: u32 = env::var("PGMAXCONNECTIONS")
            .map(|v| v.parse())
            .unwrap_or(Ok(5))?;
        let max_lifetime: u64 = env::var("PGMAXLIFETIME")
            .map(|v| v.parse())
            .unwrap_or(Ok(30 /* 30 mins */))?;

        let pgurl =
            env::var("PGURL").unwrap_or_else(|_| format!("postgres://localhost/{}", service_name));

        let mut connect_opts: PgConnectOptions = pgurl.parse()?;
        connect_opts.log_statements(log::LevelFilter::Debug);

        let pg_pool = PgPoolOptions::new()
            .max_connections(max_connections)
            .max_lifetime(Duration::from_secs(max_lifetime * 60 /* to seconds */))
            .connect_with(connect_opts)
            .await?;

        server.with(PostgresMiddleware::from(pg_pool));
    }

    Ok((base_server, server))
}

pub async fn start_server<State>(server: Server<Arc<State>>) -> Result<()>
where
    State: Send + Sync + 'static,
{
    #[cfg(feature = "lambda-http")]
    {
        server.listen(LambdaListener::new()).await?;
    }
    #[cfg(not(feature = "lambda-http"))]
    {
        let port: u16 = env::var("PORT").map(|v| v.parse()).unwrap_or(Ok(8080))?;
        let host = env::var("HOST").unwrap_or_else(|_| "127.0.0.1".to_string());

        let mut listener = server.bind((host.as_str(), port)).await?;
        for info in listener.info().iter() {
            log::info!("Server listening on {}", info);
        }
        listener.accept().await?;
    }

    // Essentially "never".
    Ok(())
}
